### PR TITLE
ledger-api-test-tool: Govern ContractIdIT test runs through a feature.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
Rather than making it optional, we can expose a feature that instructs whether we support certain kinds of contract IDs.

This works for Canton too as it defaults to "SUFFIXED".

### Changelog

- **[Ledger API Test Tool]** The "ContractIdIT" test is now run by default, using ledger feature flags to determine which test cases are run. You will need to configure the `StandaloneApiService` accordingly.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by fabiotudone-da — APPROVED** [review]
LGTM, perhaps you'll need to adjust compatibility exclusions and enable a full run?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 1]
```diff
@@ -67,3 +68,8 @@ enum CommandDeduplicationType {
   // Duplicate commands are always reported synchronously via a synchronous gRPC error on the command submission.
   SYNC_ONLY = 2;
 }
+
+message ContractIdFeatures {
+  bool v0_supported = 1;
+  bool v1_non_suffixed_supported = 2;
+}
```
Comment (meiersi-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: @SamirTalwar-DA please add comments here that are detailed enough for any of our engineers to undersand what the two features are and which ones are expected to be supported.

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: I'd expect something along one or two sentences per field and a pointer to the details which are likely present in either the test code or the LF infrastructure code.

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: Also: right now the default value of this message is such that none of the two modes are supported, which means that by default none of the corresponding tests are run. Is there the option to choose a reasonable default that works for Canton out of the box, and thereby ensure that on upgrade the test tool will always test their supported contact-ids?

Comment (remyhaemmerle-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: fyi v0 contract IDs  will be dropped with Daml 2.0, so you could drop the corresponding tests. 
language-team can take care of it, when we effectively drop the support.

Comment (remyhaemmerle-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: Probably worth add a referent to the [Contract ID spec](daml-lf/spec/contract-id.rst).

Comment (remyhaemmerle-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: Canton and KV support distinct contract IDs (non_suffixed vs suffixed), so it seems difficult to have a reasonable default.

Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: Thanks for the input @remyhaemmerle-da . The `non_suffixed vs suffixed` probably means that there should be another field or an enum that allows to represent Canton's config as well, shouldn't there?

Comment (remyhaemmerle-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:75]: @meiersi-da Good point.

We need: 
-  a features for `sandbox-on-x`, that check that both kinds of CIDs are accepted. 
 - a feature for canton, that checks that suffixed CID are accepted and non-suffixed ones are rejected.

Comment (SamirTalwar-DA) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: KV also supports both kinds of v1 CIDs. I am not sure if that's intentional, but it does.

Comment (SamirTalwar-DA) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: (You can see this in [the _ledger-on-memory_ diff](https://github.com/digital-asset/daml/pull/12454/files#diff-c3adf0e08faaddf3fdc7b9efd7bcd0fcb9bca1b5bbafe0c4c33c369fd); we used to explicitly enable the tests accordingly.)

```bazel
        "--additional=ContractIdIT:RejectV0,ContractIdIT:AcceptSuffixedV1,ContractIdIT:AcceptNonSuffixedV1",
```

Comment (SamirTalwar-DA) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: @remyhaemmerle-da, @meiersi-da, please take another look when you have a moment. I think it's clearer now.

I removed the v0 contract ID feature flag and the associated test cases.

File: compatibility/bazel_tools/testing.bzl [hunk 2]
```diff
@@ -675,6 +675,17 @@ excluded_test_tool_tests = [
             },
         ],
     },
+    {
+        # Contract ID tests are governed by feature descriptors.
+        "platform_ranges": [
+            {
+                "end": "2.0.0-snapshot.20220110.8812.1",
```
Comment (hubert-da) [inline compatibility/bazel_tools/testing.bzl:682]: Just to make sure, `ContractIdIT` is not going to be run with any `ledger-api-test-tool` version against platform versions older than the next snapshot. Is that what you want?

Comment (SamirTalwar-DA) [reply compatibility/bazel_tools/testing.bzl:682]: Yes, this reproduces the old behavior, where these tests were optional and therefore not run.

**Review by remyhaemmerle-da — COMMENTED** [review]
`ContractIdIT.scala`  looks good to me.

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto (hunk 1 above)
Comment (cocreature) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto]: This doesn’t make sense to me. We should still test that v0 contract ids are rejected especially if we’re dropping support for them. What we can drop is the test that they’re accepted.

**Review by cocreature — CHANGES_REQUESTED** [review]
Requesting changes until the v0 rejection tests are back, sorry for the confusion

File: ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala [hunk 3]
```diff
@@ -256,6 +257,10 @@ object SandboxOnXRunner {
         deduplicationType = CommandDeduplicationType.SYNC_ONLY,
         maxDeduplicationDurationEnforced = false,
       ),
+      contractIdFeatures = ContractIdFeatures.of(
+        v0 = ContractIdFeatures.ContractIdV0Support.SUPPORTED,
```
Comment (cocreature) [inline ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/SandboxOnXRunner.scala:261]: sandbox on x should not support v0 I think, only v1.

**Review by cocreature — APPROVED** [review]

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala [hunk 4]
```diff
@@ -26,171 +27,202 @@ import scala.util.{Failure, Success, Try}
...
-      "non-suffixed V1",
-      true,
-      features => features.contractIds.v1.isNonSuffixed || features.contractIds.v1.isBoth,
-      "non-suffixed V1 contract IDs are not supported",
+  List(
+    TestConfiguration(
+      description = "v0",
+      example = v0Cid,
+      accepted = true,
+      isSupported = features => features.contractIds.v0.isSupported,
```
Comment (remyhaemmerle-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:36]:
```suggestion
      isSupported = features => false,
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala [hunk 5]
```diff
@@ -26,171 +27,202 @@ import scala.util.{Failure, Success, Try}
...
-      "non-suffixed V1",
-      false,
-      features => !(features.contractIds.v1.isNonSuffixed || features.contractIds.v1.isBoth),
-      "non-suffixed V1 contract IDs are supported",
+    TestConfiguration(
+      description = "v0",
+      example = v0Cid,
+      accepted = false,
+      isSupported = features => features.contractIds.v0.isNotSupported,
+      disabledReason = "V0 contract IDs are supported",
```
Comment (remyhaemmerle-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:44]:
```suggestion
```

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto [hunk 6]
```diff
@@ -71,7 +71,13 @@ enum CommandDeduplicationType {
 
 // See `daml-lf/spec/contract-id.rst` for more information on contract ID formats.
 message ContractIdFeatures {
-  ContractIdV1Support v1 = 1;
+  ContractIdV0Support v0 = 1;
```
Comment (remyhaemmerle-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]: We do not really need a feature for V0. V0 should be rejected by all supported ledgers.
I think what @cocreature want is that we check that V0 is always rejected, but we can drop the tests that check V0 is supported.

Comment (cocreature) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]: yep

Comment (remyhaemmerle-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]:
```suggestion
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala (hunk 4 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:36]: So we never want to run this test case?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto (hunk 6 above)
Comment (SamirTalwar-DA) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]: Currently Sandbox-on-X supports v0. Should we change this?

Comment (SamirTalwar-DA) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]: (I would prefer not to change this as part of this PR.)

Comment (cocreature) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/experimental_features.proto:74]: we should but agreed, let’s do it in a separate PR.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala (hunk 5 above)
Comment (SamirTalwar-DA) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:44]: I think you meant that the other one should be deleted. This one asserts that v0 contract IDs are rejected.

Comment (remyhaemmerle-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:44]: My mistake.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:36]: We do not have any ledger that support V0. 
Sandbox classic was the only one, and it is gone.

Comment (cocreature) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ContractIdIT.scala:36]: sandbox classic and sandbox on x are both still in the repo and support it. Let’s not change that in this PR and instead separately drop support.

**Comment by SamirTalwar-DA** [comment]
Merging this but comments are still welcome.

**Pull request merged**